src/data_model/transformations.jl

```diff
@@ -472,3 +472,917 @@ function apply_voltage_angle_difference_bounds!(eng::Dict{String,<:Any}, vad::Re
+                if !haskey(data_eng["linecode"], lc_kr_id)
+                    data_eng["linecode"][lc_kr_id] = _kron_reduce_linecode(data_eng["linecode"][lc_orig_id], neutral_conductors)
+                end
+            end
+            if haskey(line, "rs")
+                _kron_reduce_linecode!(line, neutral_conductors)
+            end
+        end
+    end
+
+    # Kron-reduce each shunt if eligible
+    for (id,shunt) in data_eng["shunt"]
+        cond_is_neutral = [(shunt["bus"],t) in nbts for t in shunt["connections"]]
+        cond_keep = (!).(cond_is_neutral)
+        _apply_filter!(shunt, ["connections"], cond_keep)
+        Ys = _kron_reduce_shunt_addmittance(shunt["gs"].+im*shunt["bs"], findall(cond_is_neutral))
+        shunt["gs"] = real.(Ys)
+        shunt["bs"] = imag.(Ys)
+    end
+
+    # Kron-reduce each switch if eligible
+    for (id,switch) in data_eng["switch"]
+        doubly_grounded = [(switch["f_bus"],t_fr) in nbts && (switch["t_bus"],t_to) in nbts for (t_fr, t_to) in zip(switch["f_connections"], switch["t_connections"])]
+        if any(doubly_grounded)
+            keep = (!).(doubly_grounded)
+            neutral_conductors = findall(doubly_grounded)
+            _apply_filter!(switch, ["f_connections", "t_connections", "cm_ub", "cm_ub_b", "cm_ub_c"], keep)
+            if haskey(switch, "linecode")
+                suffix = "_kr_"*join(findall(doubly_grounded), ".")
+                lc_orig_id = switch["linecode"]
+                lc_kr_id = lc_orig_id*suffix
+                @assert !(lc_kr_id in orig_lc_ids) "Kron-reduced linecode naming clashes with original linecode names."
+                line["linecode"] = lc_kr_id
+                if !haskey(data_eng["linecode"], lc_kr_id)
+                    data_eng["linecode"][lc_kr_id] = _kron_reduce_linecode(data_eng["linecode"][lc_orig_id], neutral_conductors)
+                end
+            end
+            if haskey(switch, "rs")
+                Zs = _kron_reduce_series_impedance(switch["rs"].+im*switch["xs"], findall(doubly_grounded))
+                switch["rs"] = real.(Zs)
+                switch["xs"] = imag.(Zs)
+            end
+        end
+    end
+
+    # remove unconnected terminals (likely to be caused by earlier Kron-reductions)
+    remove_unconnected_terminals!(data_eng)
+
+    # ground remaining neutral terminals
+    remaining_bts = [(b, t) for (b,bus) in data_eng["bus"] for t in bus["terminals"]]
+    for (b,t) in intersect(nbts, remaining_bts)
+        bus = data_eng["bus"][b]
+        perfectly_grounded = bus["grounded"][iszero.(bus["rg"].+im*bus["xg"])]
+        if !(t in perfectly_grounded)
+            push!(bus["grounded"], t)
+            push!(bus["rg"], 0.0)
+            push!(bus["xg"], 0.0)
+        end
+    end
+
+    # update 'conductor_ids' property
+    find_conductor_ids!(data_eng)
+
+    return data_eng
+end
+
+
+"Remove all terminals which are unconnected (not considering a grounding as a connection)."
```

Comment:
For exported functions, we should follow the new docstring pattern where the function is given (indented) first, see, e.g. transform_data_model function as an example.